src/devices/RotaryEncoder/RotaryEncoder.cs

```diff
@@ -0,0 +1,190 @@
+        public RotaryEncoder(int pinA, int pinB) : this(new GpioController(), pinA, pinB) { }
+
+        /// <summary>
+        /// Calculate the amount of acceleration to be applied to the increment of the rotary encoder.
+        /// </summary>
+        /// <remarks>
+        /// This uses a straight line function output = input * AccelerationSlope + Acceleration offset but can be overridden 
+        /// to perform different algorithms.
+        /// </remarks>
+        /// <param name="pulseWidthMilliSeconds">The amount of time elapsed since the last data pulse from the encoder in milliseconds.</param>
+        /// <returns>A value that can be used to apply acceleration to the rotary encoder.</returns>
+        protected virtual int Acceleration(int pulseWidthMilliSeconds)
+        {
+            // apply a straight line line function to the pulseWidth to determine the acceleration but clamp the lower value to 1
+            return Math.Max(1, (int)(pulseWidthMilliSeconds * AccelerationSlope + AccelerationOffset));
+        }
+
+        /// <summary>
+        /// Modify the current value on receipt of a pulse from the rotary encoder.
+        /// </summary>
+        /// <param name="blnUp">When true then the value should be incremented otherwise it should be decremented.</param>
+        /// <param name="acceleration">A value that indicates how much acceleration should be applied to the pulse.</param>
+        protected virtual void OnPulse(bool blnUp, int acceleration)
+        {
+            // calculate how much to change the value by
+            dynamic valueChange = (blnUp ? (dynamic)_pulseIncrement : -(dynamic)_pulseIncrement) * acceleration;
+
+            // set the value to the new value clamped by the maximum and minumum of the range.
+            Value = Math.Max(Math.Min(Value + valueChange, (dynamic) _rangeMax), (dynamic) _rangeMin);
+
+            // fire an event if an event handler has been attached
+            if (ValueChanged != null)
+            {
+                ValueChanged.Invoke(this, new RotaryEncoderEventArgs<T>(Value));
+            }
+        }
+
+        /// <summary>
+        /// Initialize a RotaryEncoder
+        /// </summary>
+        /// <param name="controller">GpioController that hosts Pins A and B.</param>
+        /// <param name="pinA">Pin A that is connected to the rotary encoder. Sometimes called clk</param>
+        /// <param name="pinB">Pin B that is connected to the rotary encoder. Sometimes called data</param>
+        /// <param name="pulseIncrement">The amount that the value increases or decreases on each pulse from the rotary encoder</param>
+        /// <param name="rangeMin">Minimum value permitted. The value is clamped to this.</param>
+        /// <param name="rangeMax">Maximum value permitted. The value is clamped to this.</param>
+        private void Initialize(GpioController controller, int pinA, int pinB, T pulseIncrement, T rangeMin, T rangeMax)
```

Comment:
would it makes sense to instead have `number of pulses per rotation` then have properties:

- PulseCountPerRotation
- PulseCount
- NumberOfRotations
- NumberOfRotationsPerSecond (equivalent to current acceleration)

all rotary encoders I find on Amazon display `pulses/rotation` so it would be nice if you could just type it in the constructor

Reply:
The design was to allow the encoder to provide values that reflected the use to which it was being put. I have used the FM radio scale in the demo 88.0 to 108.0 in 0.1 Mhz steps. I believe that building the scaling and the maths into the control makes sense. Another application could be setting the time for a digital clock with perhaps a with a TimeSpan as T and overriding the OnPulse to add or subtract a certain number of seconds.

I am a bit confused why you are suggesting properties related to rotation as most uses of its use case wouldn't care how far in angular terms it has been turned (unless it's a shaft encoder) . 

I don't think double is a good choice as add 0.1D to 0 three times gives me 0.300000000000004 and this could cause issues and this is why I chose to implement this as a generic as you could use whatever type was appropriate. You are right that the users could do the maths themselves but I am trying not to constrain them to types

Also the properties PulseCount NumberOfRotations do not make any sense to me as they would have to be relative to something would it be the last time the property was read or something else.

Perhaps we have a different view about where these encoders might me used.

Reply:
@Frankenslag I don't disagree that builtin math is useful there but my point is that before you add something more advanced you should first provide the basic functionality which is what the basic parameters are.

Let's say I buy rotary encoder now - the only parameter I can find on the website is P/R (pulses per rotation) - this is what I'd expect to pass in to the constructor.

> Why have anything related with rotations?
One of the usages of rotary encoder is to connect it with motor to see how fast it is rotating.
Other reason is it's really easy to think about - you rotate once, you expect 1 to be somewhere in the properties.

> Also the properties PulseCount NumberOfRotations do not make any sense to me as they would have to be relative to something would it be the last time the property was read or something else.
You can ask identical question to your case - what would be the initial number? I'd say 0 is probably expected, optionally you could provide your own number or have a way to reset it

> I don't think double is a good choice as add 0.1D to 0 three times gives me 0.300000000000004

You have identical problem with your current implementation, also if you calculate in terms of pulses/rotations and have pulsesPerRotations then (untested):

```csharp
value = ((double)pulses / pulsesPerRotation) * (maxRange - minRange) + minRange;
```

will add as accurately as it is possible and increment by 0.1 doesn't exist since you calculate on integers (long or int)

My point is to have at minimum `Pulses` and `PulsesPerRotation` exposed, anything else can be calculated from that. If you expose only something like `Value` then it actually makes it harder to go the other way

Reply:
@krwq
I take your point with rotation and pulses and rotations as properties. Perhaps we should have a base class that works that way and create a subclass to deal with values and ranges. What do you think ?

Reply:
@krwq @hugener
I have created a base class called `QuadratureRotaryEncoder` which I hope delivers the rotation properties. I have then created a subclass for a `ScaledQuadratureEncoder` to replace the previous `RotaryEncoder`. I hope that this naming describes things a bit better.

I have also added the ability to define which Gpio events the encoder is listening for.

I have updated the PR to reflect this.